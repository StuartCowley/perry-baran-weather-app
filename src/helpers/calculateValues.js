const calcMean = (array, param, decimals = 1) => {
  return (
    array.reduce((prev, curr) => {
      return prev + curr[param];
    }, 0) / array.length
  ).toFixed(decimals);
};

const calcMax = (array, param, decimals = 1) => {
  const values = array.map((item) => {
    return item[param];
  });
  return Math.max(...values).toFixed(decimals);
};

const calcMin = (array, param, decimals = 1) => {
  const values = array.map((item) => {
    return item[param];
  });
  return Math.min(...values).toFixed(decimals);
};

export { calcMean, calcMax, calcMin };
