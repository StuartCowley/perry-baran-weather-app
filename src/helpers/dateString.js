const dateString = (dateTime) => {
  return new Date(dateTime * 1000).toDateString();
};

export default dateString;
