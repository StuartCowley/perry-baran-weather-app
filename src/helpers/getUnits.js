const getUnits = (units) => {
  let selectedUnits;
  if (typeof units === "string") selectedUnits = units.toLowerCase();
  switch (selectedUnits) {
    case "imperial":
      return {
        tempUnits: "°F",
        speedUnits: "mph",
      };
    case "metric":
      return {
        tempUnits: "°C",
        speedUnits: "m/s",
      };
    case "standard":
    default:
      return {
        tempUnits: "K",
        speedUnits: "m/s",
      };
  }
};

export default getUnits;
