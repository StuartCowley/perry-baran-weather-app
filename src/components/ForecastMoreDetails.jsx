import React from "react";
import "../styles/ForecastDetails.css";
import PropTypes from "prop-types";
import ForecastTrihourlySummaries from "./ForecastTrihourlySummaries";
import ForecastBreakdown from "./ForecastBreakdown";
import { calcMean, calcMax, calcMin } from "../helpers/calculateValues";
import dateString from "../helpers/dateString";

function ForecastMoreDetails({ forecasts }) {
  const [{ dateTime, icon, weather }] = forecasts;
  const date = dateString(dateTime);
  const temp = calcMean(forecasts, "temp");
  const minTemp = calcMin(forecasts, "minTemp");
  const maxTemp = calcMax(forecasts, "maxTemp");
  const humidity = calcMean(forecasts, "humidity");
  const windSpeed = calcMean(forecasts, "windSpeed");

  return (
    <div className="forecast-details">
      <h2>{date}</h2>
      <ForecastBreakdown
        icon={icon}
        temp={temp}
        minTemp={minTemp}
        maxTemp={maxTemp}
        humidity={humidity}
        windSpeed={windSpeed}
        weather={weather}
      />
      <ForecastTrihourlySummaries forecasts={forecasts} />
    </div>
  );
}

ForecastMoreDetails.propTypes = {
  forecasts: PropTypes.arrayOf(
    PropTypes.shape({
      dateTime: PropTypes.number,
      dateTimeTxt: PropTypes.string,
      icon: PropTypes.string,
      temp: PropTypes.number,
      maxTemp: PropTypes.number,
      minTemp: PropTypes.number,
      humidity: PropTypes.number,
      weather: PropTypes.string,
      windSpeed: PropTypes.number,
    })
  ).isRequired,
};

export default ForecastMoreDetails;
