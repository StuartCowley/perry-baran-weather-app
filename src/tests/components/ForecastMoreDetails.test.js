import React from "react";
import { render } from "@testing-library/react";
import ForecastMoreDetails from "../../components/ForecastMoreDetails";
import dateString from "../../helpers/dateString";
import { calcMax, calcMean, calcMin } from "../../helpers/calculateValues";

describe("ForecastMoreDetails", () => {
  const forecasts = [
    {
      dateTime: 1667487600,
      dateTimeTxt: "2022-11-03 15:00:00",
      icon: "200",
      temp: 25,
      maxTemp: 22,
      minTemp: 27,
      humdidity: 29,
      weather: "light rain",
      windSpeed: 100,
    },
    {
      dateTime: 2667554800,
      dateTimeTxt: "2022-11-04 12:00:00",
      icon: "500",
      temp: 35,
      maxTemp: 32,
      minTemp: 37,
      humdidity: 39,
      weather: "broken clouds",
      windSpeed: 90,
    },
  ];
  let screen;

  beforeEach(() => {
    screen = render(<ForecastMoreDetails forecasts={forecasts} />);
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("data is properly manipulated", () => {
    const { getByText } = screen;
    const date = dateString(forecasts[0].dateTime);
    const temp = calcMean(forecasts, "temp");
    const minTemp = calcMin(forecasts, "minTemp");
    const maxTemp = calcMax(forecasts, "maxTemp");
    const humidity = calcMean(forecasts, "humidity");
    const windSpeed = calcMean(forecasts, "windSpeed");

    expect(getByText(date)).toBeInstanceOf(HTMLHeadingElement);
    expect(getByText(new RegExp(temp))).toBeInTheDocument();
    expect(getByText(new RegExp(minTemp))).toBeInTheDocument();
    expect(getByText(new RegExp(maxTemp))).toBeInTheDocument();
    expect(getByText(new RegExp(humidity))).toBeInTheDocument();
    expect(getByText(new RegExp(windSpeed))).toBeInTheDocument();
  });
});
