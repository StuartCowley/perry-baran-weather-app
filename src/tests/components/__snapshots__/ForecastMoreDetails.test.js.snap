// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastMoreDetails snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-details"
  >
    <h2>
      Thu Nov 03 2022
    </h2>
    <div
      class="forecast-breakdown"
    >
      <div
        class="forecast-breakdown__right"
      >
        <div
          data-testid="forecast-icon"
        >
          <i
            class="wi wi-day-storm-showers"
            name="owm"
          />
        </div>
        <p>
          light rain
        </p>
      </div>
      <div
        class="forecast-breakdown__left"
      >
        <p
          class="forecast-breakdown__temp"
        >
          30.0
          <sup>
            °C
          </sup>
        </p>
        <div>
          <h3>
            Min: 
          </h3>
          <p>
            27.0°C
          </p>
        </div>
        <div>
          <h3>
            Max: 
          </h3>
          <p>
            32.0°C
          </p>
        </div>
        <div>
          <h3>
            Humidity: 
          </h3>
          <p>
            NaN%
          </p>
        </div>
        <div>
          <h3>
            Wind Speed: 
          </h3>
          <p>
            95.0m/s
          </p>
        </div>
      </div>
    </div>
    <div
      class="forecast-trihourly-summaries"
    >
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          00:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          03:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          06:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          09:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          12:00
        </p>
        <div
          class="forecast-trihourly-summary__icon"
          data-testid="forecast-icon"
        >
          <i
            class="wi wi-day-rain"
            name="owm"
          />
        </div>
        <p>
          broken 
          <br />
          clouds
        </p>
        <p>
          35.0°C
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          15:00
        </p>
        <div
          class="forecast-trihourly-summary__icon"
          data-testid="forecast-icon"
        >
          <i
            class="wi wi-day-storm-showers"
            name="owm"
          />
        </div>
        <p>
          light 
          <br />
          rain
        </p>
        <p>
          25.0°C
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          18:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
      <div
        class="forecast-trihourly-summary"
        data-testid="forecast-trihourly-summary"
      >
        <p>
          21:00
        </p>
        <p>
          No 
          <br />
          Data
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
